src/compiler/checker.ts

```diff
@@ -30855,7 +30861,21 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             const jsxOpeningLikeNode = node ;
             const sig = getResolvedSignature(jsxOpeningLikeNode);
             checkDeprecatedSignature(sig, node);
-            checkJsxReturnAssignableToAppropriateBound(getJsxReferenceKind(jsxOpeningLikeNode), getReturnTypeOfSignature(sig), jsxOpeningLikeNode);
+
+            const elementTypeConstraint = getJsxElementTypeTypeAt(jsxOpeningLikeNode);
+            if (elementTypeConstraint !== undefined) {
+                const tagType = isJsxIntrinsicIdentifier(jsxOpeningLikeNode.tagName)
+                    ? // TODO: Should this be a literal that library authors could potentially check against?

```

Comment:
I guess this would only be for library authors providing a JSX factory of some sort, right? I don't have a strong opinion on this.

Reply:
Yes, it should be a literal type. The literal type will be assignable to `string`, and forwarding the literal type lets library authors filter the allowed types to only actually existing ones.

Reply:
Ugh, in fact it *has* to be a literal if it's called `ElementType` type because Preact already declares `ElementType`

https://github.com/preactjs/preact/blob/90042a1a012855ef975309485534595705c777c5/src/jsx.d.ts#L34-L40

Which is why the following tests break in DefinitelyTyped.

https://github.com/microsoft/TypeScript/pull/51328#issuecomment-1504224927

Reply:
Also, given that `ElementType` is generic for them, wouldn't we have to instantiate with default arguments? Do you know if that's necessary @weswigham?

Reply:
I kept it as an abstract string since the checker later checked against known host element types in `JSX.IntrinsicElements` anyway and still kept the old error message (see https://github.com/microsoft/TypeScript/pull/51328#discussion_r1106058919).

So unknown host elements are still rejected and I think we still need `JSX.IntrinsicElements` anyway for props assignability. 


I added a test to verify that unknown host element types are still rejected and then switched to `getStringLiteralType(tagName.escapedText as string)` which didn't change any behavior as far as I can tell.

Reply:
I don't *think* you can do `tagName.escapedText as string` because the escaping won't correctly match up in a case like:

```ts
namespace JSX {
    // ...
    type ElementType = "__";
}

// ...

function __() {
    return <div />
}

let x = <__></__>;
```